src/test/java/org/elasticsearch/cluster/ClusterInfoServiceTests.java

```diff
@@ -0,0 +1,259 @@
+        assertThat("some shard sizes are populated", info.getShardSizes().size(), greaterThan(0));
+
+
+        MockTransportService mockTransportService = (MockTransportService) internalCluster().getInstance(TransportService.class, internalTestCluster.getMasterName());
+
+        final AtomicBoolean timeout = new AtomicBoolean(false);
+        final Set<String> blockedActions = ImmutableSet.of(NodesStatsAction.NAME, NodesStatsAction.NAME + "[n]", IndicesStatsAction.NAME, IndicesStatsAction.NAME + "[s]");
+        for (DiscoveryNode node : internalTestCluster.clusterService().state().getNodes()) {
+            mockTransportService.addDelegate(node, new MockTransportService.DelegateTransport(mockTransportService.original()) {
+                @Override
+                public void sendRequest(DiscoveryNode node, long requestId, String action, TransportRequest request, TransportRequestOptions options) throws IOException, TransportException {
+                    if (blockedActions.contains(action)) {
+                        if (timeout.get()) {
+                            logger.info("dropping [{}] to [{}]", action, node);
+                            return;
+                        }
+                    }
+                    super.sendRequest(node, requestId, action, request, options);
+                }
+            });
+        }
+
+        // timeouts shouldn't clear the info
+        timeout.set(true);
+        listener.reset();
+        infoService.updateOnce();
+        info = listener.get();
+        assertNotNull("info should not be null", info);
+        // node info will time out both on the request level on the count down latch. this means
+        // it is likely to update the node disk usage based on the one response that came be from local
+        // node.
+        assertThat(info.getNodeDiskUsages().size(), greaterThanOrEqualTo(1));
+        // indices is guaranteed to time out on the latch, not updating anything.
+        assertThat(info.getShardSizes().size(), greaterThan(1));
+
+        // now we cause an exception
+        timeout.set(false);
+        ActionFilters actionFilters = internalTestCluster.getInstance(ActionFilters.class, internalTestCluster.getMasterName());
+        BlockingActionFilter blockingActionFilter = null;
+        for (ActionFilter filter : actionFilters.filters()) {
+            if (filter instanceof BlockingActionFilter) {
+                blockingActionFilter = (BlockingActionFilter) filter;
+                break;
+            }
+        }
+
+        assertNotNull("failed to find BlockingActionFilter", blockingActionFilter);
+        blockingActionFilter.blockActions(blockedActions.toArray(Strings.EMPTY_ARRAY));
+        listener.reset();
+        infoService.updateOnce();
+        info = listener.get();
+        assertNotNull("info should not be null", info);
+        assertThat(info.getNodeDiskUsages().size(), equalTo(0));
+        assertThat(info.getShardSizes().size(), equalTo(0));
```

Comment:
Can you add an additional check after this where after the (hopefully intermittent) issue has gone away, the sizes are correctly set?